### PR TITLE
FIX-#3702: add proper check for none-slice at `df.__getitem__`

### LINKED ISSUE #3702: `DataFrame.__getitem__` does not respect the `step` attribute of a slice argument
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): 8d67cd3a467d1eb2262898f6f5e2cf159c27d7b0
- **Python version**: Any
- **Code we can use to reproduce**:
```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2, 3, 4], "b": [3, 4, 5, 6]})

print(f"Modin result:\n{df[::2]}\n")
print(f"Pandas result:\n{df._to_pandas()[::2]}\n")
```

Output:
```
Modin result:
   a  b
0  1  3
1  2  4
2  3  5
3  4  6

Pandas result:
   a  b
0  1  3
2  3  5
```

### Describe the problem
The problem is in this incorrect check for None-slice, it does not respect the `step` parameter:
https://github.com/modin-project/modin/blob/8d67cd3a467d1eb2262898f6f5e2cf159c27d7b0/modin/pandas/base.py#L2999-L3000

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

Adds a unified check of whether a slice indexer grabs the whole sequence (`is_full_grab_slice`) and replaces an incorrect full-grab check at the `df.__getitem__` with the call of the new function.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3702 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/utils.py [hunk 1]
```diff
@@ -113,6 +113,33 @@ def is_scalar(obj):
...
+
+    Parameters
+    ----------
+    slc : slice
+        Slice object to check.
+    sequence_len : int, optional
+        Length of the sequence to index with the passed `slc`.
+        If not specified the function won't consider `slc` to
+        be a full-grab if its ``.stop`` attribute is equal or
+        greater than the sequence length.
```
Comment (vnlitvinov) [inline modin/pandas/utils.py:128]: wait, how the function can know sequence length if it's `None` to compare `slc.stop` to it?.. something's wrong with this docstring, pls fix

**Review by vnlitvinov — CHANGES_REQUESTED** [review]
Fix the docstring, otherwise LGTM

File: modin/pandas/utils.py (hunk 1 above)
Comment (dchigarev) [reply modin/pandas/utils.py]: Changed to the following:
"If not specified the function won't be able to check whether ``slc.stop`` is equal or greater than the sequence length to consider `slc` to be a full-grab, and so, only slices with ``.stop is None`` are considered to be a full-grab."

**Review by vnlitvinov — DISMISSED** [review]
LGTM, thanks!

File: modin/pandas/utils.py [hunk 2]
```diff
@@ -113,6 +113,34 @@ def is_scalar(obj):
...
+        If not specified the function won't be able to check whether
+        ``slc.stop`` is equal or greater than the sequence length to
+        consider `slc` to be a full-grab, and so, only slices with
+        ``.stop is None`` are considered to be a full-grab.
+
+    Returns
+    -------
+    bool
+    """
+    return (
```
Comment (devin-petersohn) [inline modin/pandas/utils.py:135]:
```suggestion
    assert isinstance(slc, slice), "slice object required"
    return (
```

**Review by devin-petersohn — COMMENTED** [review]
Looks great, just one defensive coding suggestion. thanks @dchigarev!

**Review by devin-petersohn — APPROVED** [review]
LGTM! Will let @vnlitvinov take another look.

**Review by vnlitvinov — APPROVED** [review]
LGTM, thanks!

**Pull request merged**